### PR TITLE
GH-43221: [C++][Parquet] Refactor parquet::encryption::AesEncryptor to use unique_ptr

### PR DESCRIPTION
**Pull request by mapleFU**
### Rationale for this change

See https://github.com/apache/arrow/issues/43221

### What changes are included in this PR?

Change raw-pointer to unique_ptr

### Are these changes tested?

Covered by existing

### Are there any user-facing changes?

Maybe change user interface

* GitHub Issue: #43221

### CODE REVIEW

File: cpp/src/parquet/encryption/internal_file_encryptor.cc
Comment (pitrou) [inline cpp/src/parquet/encryption/internal_file_encryptor.cc:154]: Unless I'm mistaken, `all_encryptors_` is never popped from when an encryptor is deallocated. Could we fix this?

Comment (mapleFU) [reply cpp/src/parquet/encryption/internal_file_encryptor.cc:154]: seems that is owned by `meta_encryptor_` and `data_encryptor_`?

Comment (mapleFU) [reply cpp/src/parquet/encryption/internal_file_encryptor.cc:154]: All also we can remove `all_encryptors_`, and do:

```
void InternalFileEncryptor::WipeOutEncryptionKeys() {
  properties_->WipeOutEncryptionKeys();

  for (auto const& i : meta_encryptor_) {
    if (i != nullptr) i->WipeOut();
  }
  for (auto const& i : data_encryptor_) {
    if (i != nullptr) i->WipeOut();
  }
}
```

Comment (pitrou) [reply cpp/src/parquet/encryption/internal_file_encryptor.cc:154]: Sounds ok to me, though why the nullptr check?

Comment (mapleFU) [reply cpp/src/parquet/encryption/internal_file_encryptor.cc:154]: `GetMetaAesEncryptor` does a nullable check, I wonder would we face the case that we didn't use it but it being WipeOut

Comment (mapleFU) [reply cpp/src/parquet/encryption/internal_file_encryptor.cc:154]:
```
  // Key must be 16, 24 or 32 bytes in length. Thus there could be up to three
  // types of meta_encryptors and data_encryptors.
  std::unique_ptr<encryption::AesEncryptor> meta_encryptor_[3];
  std::unique_ptr<encryption::AesEncryptor> data_encryptor_[3];
```

And from the comment it might be nullptr

Comment (pitrou) [reply cpp/src/parquet/encryption/internal_file_encryptor.cc:154]: I see, SGTM

File: cpp/src/parquet/metadata.cc
Comment (mapleFU) [inline cpp/src/parquet/metadata.cc:656]: @adamreeve @wgtmac I'm not so familiar with logic here but `AesEncryptor::Make` might return nullptr if aes not enable, so should we add this check?

Comment (mapleFU) [reply cpp/src/parquet/metadata.cc:656]: exception is better than failed and core dump here?

Comment (adamreeve) [reply cpp/src/parquet/metadata.cc:656]: Yes I think this should be checked and I agree an exception is nicer than abort.

It looks like other places that construct an AesEncryptor don't check for null either though, eg. the encryptors returned from `InternalFileEncryptor::GetMetaAesEncryptor` and `InternalFileEncryptor::GetDataAesEncryptor` appear to be used without checking for null. Maybe it would be better to change `Make` to return `Result<std::unique_ptr<AesEncryptor>>` to enforce checking for failure, or possibly it makes sense to just change the `_nossl` versions of `Make` to throw an exception rather than return null?

Comment (mapleFU) [reply cpp/src/parquet/metadata.cc:656]: > Maybe it would be better to change Make to return Result<std::unique_ptr<AesEncryptor>> to enforce checking for failure, or possibly it makes sense to just change the _nossl versions of Make to throw an exception rather than return null, as I can't see anywhere that handles a null value already?

Currently I'm not familiar with encryption enough, so I just put the checking here🤔

> AesDecryptor::Make also seems to have the same problem.

Yeah I'll update it here

File: cpp/src/parquet/encryption/internal_file_encryptor.cc
Comment (pitrou) [inline cpp/src/parquet/encryption/internal_file_encryptor.cc:56]: If we ensure that `AesEncryptor::Make` and `AesDecryptor::Make` always return a non-null value, we don't need those checks.

File: cpp/src/parquet/metadata.cc
Comment (pitrou) [inline cpp/src/parquet/metadata.cc:668]: What's the point of this? `aes_encryptor` is just a local variable.

**Review by pitrou — CHANGES_REQUESTED** [review]

File: cpp/src/parquet/metadata.cc
Comment (mapleFU) [reply cpp/src/parquet/metadata.cc]: This is to keep same behavior as before, but actually it's not need

File: cpp/src/parquet/encryption/encryption_internal_nossl.cc
Comment (mapleFU) [reply cpp/src/parquet/encryption/encryption_internal_nossl.cc:78]: done

File: cpp/src/parquet/encryption/internal_file_encryptor.cc
Comment (mapleFU) [reply cpp/src/parquet/encryption/internal_file_encryptor.cc:57]:
```
  // Key must be 16, 24 or 32 bytes in length. Thus there could be up to three
  // types of meta_encryptors and data_encryptors.
  std::unique_ptr<encryption::AesEncryptor> meta_encryptor_[3];
  std::unique_ptr<encryption::AesEncryptor> data_encryptor_[3];
```

This is as said before, since it might only has one?

Comment (pitrou) [reply cpp/src/parquet/encryption/internal_file_encryptor.cc:57]: Ah, sorry, I had forgotten about this. Nevermind.

File: cpp/src/parquet/metadata.cc
Comment (pitrou) [inline cpp/src/parquet/metadata.cc:657]: I think we can remove this check now that we know `AesEncryptor::Make` will never return a null pointer?

File: cpp/src/parquet/encryption/encryption_internal_nossl.cc
Comment (pitrou) [inline cpp/src/parquet/encryption/encryption_internal_nossl.cc:77]: Can you also make the same change for `AesDecryptor::Make`?

**Review by pitrou — CHANGES_REQUESTED** [review]

**Comment by pitrou** [comment]
@github-actions crossbow submit -g cpp -g python

**Review by pitrou — APPROVED** [review]
Thanks for the update! LGTM, let's just wait for the Crossbow CI jobs before merging.

**Comment by mapleFU** [comment]
The emscript failed in csv test because no resource:

```
[ RUN      ] NullColumnDecoderTest.Threaded
unknown file: Failure
C++ exception with description "thread constructor failed: Resource temporarily unavailable" thrown in the test body.
[  FAILED  ] NullColumnDecoderTest.Threaded (0 ms)
```

**Pull request merged**
